rcl/test/rcl/test_subscription.cpp

```diff
@@ -0,0 +1,235 @@
+  void SetUp()
+  {
+    stop_memory_checking();
+    auto id = rmw_get_implementation_identifier();
+    fprintf(stderr, "RMW MIDDLEWARE ID:          %s(%p)\n", id, id);
+    auto ts = ROSIDL_GET_MSG_TYPE_SUPPORT(std_msgs, Int64);
+    ASSERT_NE(nullptr, ts);
+    id = ts->typesupport_identifier;
+    fprintf(stderr, "TYPE SUPPORT MIDDLEWARE ID: %s(%p)\n", id, id);
+    rcl_ret_t ret;
+    ret = rcl_init(0, nullptr, rcl_get_default_allocator());
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+    this->node_ptr = new rcl_node_t;
+    *this->node_ptr = rcl_get_zero_initialized_node();
+    const char * name = "node_name";
+    rcl_node_options_t node_options = rcl_node_get_default_options();
+    ret = rcl_node_init(this->node_ptr, name, &node_options);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+    set_on_unexpected_malloc_callback([]() {ASSERT_FALSE(true) << "UNEXPECTED MALLOC";});
+    set_on_unexpected_realloc_callback([]() {ASSERT_FALSE(true) << "UNEXPECTED REALLOC";});
+    set_on_unexpected_free_callback([]() {ASSERT_FALSE(true) << "UNEXPECTED FREE";});
+    start_memory_checking();
+  }
+
+  void TearDown()
+  {
+    assert_no_malloc_end();
+    assert_no_realloc_end();
+    assert_no_free_end();
+    stop_memory_checking();
+    set_on_unexpected_malloc_callback(nullptr);
+    set_on_unexpected_realloc_callback(nullptr);
+    set_on_unexpected_free_callback(nullptr);
+    rcl_ret_t ret = rcl_node_fini(this->node_ptr);
+    delete this->node_ptr;
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+    ret = rcl_shutdown();
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+  }
+};
+
+void
+wait_for_subscription_to_be_ready(
+  rcl_subscription_t * subscription,
+  size_t retries,
+  int64_t period_ms,
+  bool & success)
+{
+  rcl_wait_set_t wait_set = rcl_get_zero_initialized_wait_set();
+  rcl_ret_t ret = rcl_wait_set_init(&wait_set, 1, 0, 0, rcl_get_default_allocator());
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+  auto wait_set_exit = make_scope_exit([&wait_set]() {
+    stop_memory_checking();
+    rcl_ret_t ret = rcl_wait_set_fini(&wait_set);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+  });
+  size_t iteration = 0;
+  do {
+    iteration++;
+    ret = rcl_wait_set_clear_subscriptions(&wait_set);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+    ret = rcl_wait_set_add_subscription(&wait_set, subscription);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+    ret = rcl_wait(&wait_set, RCL_MS_TO_NS(period_ms));
+    if (ret == RCL_RET_TIMEOUT) {
+      continue;
+    }
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+    for (size_t i = 0; i < wait_set.size_of_subscriptions; ++i) {
+      if (wait_set.subscriptions[i] && wait_set.subscriptions[i] == subscription) {
+        success = true;
+        return;
+      }
+    }
+  } while (iteration < retries);
```

Comment:
The semantic of `retries` seems weird. If passed `1` the loop will only be executed once since `iteration` is set to `1` in the first line of the loop. I would expect *one retry*?